src/main/java/com/hivemq/mqtt/message/pool/FreePacketIdRanges.java

```diff
@@ -0,0 +1,185 @@
+     *
+     * @param id an ID that the caller attempts to take.
+     * @return the requested {@param id} if it is available in one of the {@link Range} intervals or otherwise some
+     *         other free ID.
+     */
+    public int takeIfAvailable(final int id) throws NoMessageIdAvailableException {
+        Preconditions.checkArgument(id >= MIN_ALLOWED_MQTT_PACKET_ID && id <= MAX_ALLOWED_MQTT_PACKET_ID,
+                "Attempting to take an ID %s that is outside the valid packet IDs range.",
+                id);
+
+        Range current = rootRange;
+        Range prev = null;
+
+        while (current != null) {
+            if (id >= current.start && id < current.end) {
+
+                final int prevCurStart = current.start;
+                current.start = id + 1;
+                final Range lowerRange = prevCurStart == id ? null : new Range(prevCurStart, id, current);
+
+                if (lowerRange != null) {
+                    if (prev != null) {
+                        prev.next = lowerRange;
+                    } else {
+                        rootRange = lowerRange;
+                    }
+                }
+
+                return id;
+            }
+
+            prev = current;
+            current = current.next;
+        }
+
+        return takeNextId();
+    }
+
+    /**
+     * Returns the {@param id} into one of the ranges of free IDs.
+     *
+     * @param id an ID that the caller attempts to return (to free).
+     */
+    public void returnId(final int id) {
+        Preconditions.checkArgument(id >= MIN_ALLOWED_MQTT_PACKET_ID && id <= MAX_ALLOWED_MQTT_PACKET_ID,
+                "Attempting to return an ID %s that is outside the valid packet IDs range.",
+                id);
+
+        Range current = rootRange;
+        if (id < current.start - 1) { // at least one element is between the returned and the next range
+            rootRange = new Range(id, id + 1, current);
+            return;
+        }
+        Range prev = current;
+        current = returnId(current, id);
+        while (current != null) {
+            if (id < current.start - 1) {
+                prev.next = new Range(id, id + 1, current);
+                return;
+            }
+            prev = current;
+            current = returnId(current, id);
+        }
+    }
+
+    private @Nullable Range returnId(final @NotNull Range range, final int id) throws IllegalStateException {
+        if (id == range.start - 1) { // if the returned element is directly adjacent to the range (from below)
+            range.start = id;
+            return null;
+        }
+
+        if (id < range.end) { // the returned element is within the range, i.e. it has been freed already
+            return null;
+        }
+
+        final Range next = range.next;
+        if (id == range.end) {
+            Preconditions.checkState(next != null, "The id is greater than maxId. This must not happen and is a bug.");
```

Comment:
We do this precondition in either case of the if statement. It should just move before it.